### ISSUE TITLE
Detaching gradients instead of using no_grad in spectral_norm

### ISSUE THREAD

**Issue by ferrine** [pytorch/pytorch#10004]
Looking at [this](https://github.com/pytorch/pytorch/blob/3609977d7f4629f75dd1f8d904ddd5b52388124f/torch/nn/utils/spectral_norm.py#L53) code I've noticed that there is an easy way to make `u`, `v` differentiable wrt `W` (it is useful as normalizing constant `sigma` depends on `u`, `v` and `u`, `v` depend on `W`). 
If we `detach` starting point of power iteration `u` we avoid cumulating gradients problem while allowing gradients pass through power iteration. As it adds some computational overhead it can be optional.

**Comment by SsnL** [comment]
The original Spectral Norm paper doesn't backprop through the left and right singular vectors, so didn't we. It's also unclear why detaching before the loop was reasonable because the _u vector is cached and reused.

**Comment by ferrine** [comment]
According to https://discuss.pytorch.org/t/disconnected-gradient-in-pytorch/6341/5
detach will help to avoid a leak.

**Comment by SsnL** [comment]
no_grad stops history tracking too.

**Comment by ferrine** [comment]
I see, that was just a proposal to correctly pass gradients through power
iteration method

пн, 30 июл. 2018 г., 22:51 Tongzhou Wang <notifications@github.com>:

**Comment by SsnL** [comment]
@ferrine No worries. This is a perfectly valid thing to think about. I considered that too when we implemented SN. But _u and _v are cached at every iteration so it would be (nearly) impossible to compute exact gradient.
